federation/pkg/federatedtypes/hpa.go

```diff
@@ -0,0 +1,762 @@
+}
+
+func (a *HpaAdapter) FedUpdate(obj pkgruntime.Object) (pkgruntime.Object, error) {
+	hpa := obj.(*autoscalingv1.HorizontalPodAutoscaler)
+	return a.client.AutoscalingV1().HorizontalPodAutoscalers(hpa.Namespace).Update(hpa)
+}
+
+func (a *HpaAdapter) FedWatch(namespace string, options metav1.ListOptions) (watch.Interface, error) {
+	return a.client.AutoscalingV1().HorizontalPodAutoscalers(namespace).Watch(options)
+}
+
+func (a *HpaAdapter) ClusterCreate(client kubeclientset.Interface, obj pkgruntime.Object) (pkgruntime.Object, error) {
+	hpa := obj.(*autoscalingv1.HorizontalPodAutoscaler)
+	return client.AutoscalingV1().HorizontalPodAutoscalers(hpa.Namespace).Create(hpa)
+}
+
+func (a *HpaAdapter) ClusterDelete(client kubeclientset.Interface, nsName types.NamespacedName, options *metav1.DeleteOptions) error {
+	return client.AutoscalingV1().HorizontalPodAutoscalers(nsName.Namespace).Delete(nsName.Name, options)
+}
+
+func (a *HpaAdapter) ClusterGet(client kubeclientset.Interface, namespacedName types.NamespacedName) (pkgruntime.Object, error) {
+	return client.AutoscalingV1().HorizontalPodAutoscalers(namespacedName.Namespace).Get(namespacedName.Name, metav1.GetOptions{})
+}
+
+func (a *HpaAdapter) ClusterList(client kubeclientset.Interface, namespace string, options metav1.ListOptions) (pkgruntime.Object, error) {
+	return client.AutoscalingV1().HorizontalPodAutoscalers(namespace).List(options)
+}
+
+func (a *HpaAdapter) ClusterUpdate(client kubeclientset.Interface, obj pkgruntime.Object) (pkgruntime.Object, error) {
+	hpa := obj.(*autoscalingv1.HorizontalPodAutoscaler)
+	return client.AutoscalingV1().HorizontalPodAutoscalers(hpa.Namespace).Update(hpa)
+}
+
+func (a *HpaAdapter) ClusterWatch(client kubeclientset.Interface, namespace string, options metav1.ListOptions) (watch.Interface, error) {
+	return client.AutoscalingV1().HorizontalPodAutoscalers(namespace).Watch(options)
+}
+
+func (a *HpaAdapter) NewTestObject(namespace string) pkgruntime.Object {
+	var min int32 = 4
+	var targetCPU int32 = 70
+	return &autoscalingv1.HorizontalPodAutoscaler{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "test-hpa-",
+			Namespace:    namespace,
+		},
+		Spec: autoscalingv1.HorizontalPodAutoscalerSpec{
+			ScaleTargetRef: autoscalingv1.CrossVersionObjectReference{
+				Kind: "replicaset",
+				Name: "myrs",
+			},
+			MinReplicas:                    &min,
+			MaxReplicas:                    int32(10),
+			TargetCPUUtilizationPercentage: &targetCPU,
+		},
+	}
+}
+
+func (a *HpaAdapter) ImplementsReconcilePlugin() bool {
+	return true
+}
+
+type replicaNums struct {
+	min int32
+	max int32
+}
+
+type hpaLists struct {
+	availableMin sets.String
+	availableMax sets.String
+	noHpa        sets.String
+}
+
+func (a *HpaAdapter) ReconcileHook(fedObj pkgruntime.Object, currentObjs map[string]pkgruntime.Object) (map[string]pkgruntime.Object, error) {
+	fedHpa := fedObj.(*autoscalingv1.HorizontalPodAutoscaler)
+	requestedMin := int32(1)
```

Comment:
Actually if you see the code, this is used only in the case when this field value is not already set (Spec.MinReplicas == nil on the object), which is unlikely because if the object for whatever reason (for example user does not specify it while creating) this field is unspecified, and if the request came from valid route (creating into the api server), the default will already by set by api defaults (the code you pointed at), even if its something else then 1. However, I am still inclined at keeping this local default, in case this code is invoked from tests, which might provide an object with this field unset. I have made it a constant and provided a comment to elaborate its use.